### ISSUE TITLE
Multiple go_repository rules with same name should be an error

### ISSUE THREAD

**Issue by drigz** [bazelbuild/rules_go#749]
If the WORKSPACE contains multiple go_repository() rules with the same name but different commits, the first is silently ignored.

bazelbuild/bazel#1984 discusses rule redefinition in Bazel, where it is an error after the first `load()` statement. 

### Reproduction

```
# WORKSPACE
http_archive(
    name = "io_bazel_rules_go",
    sha256 = "f7e42a4c1f9f31abff9b2bdee6fe4db18bc373287b7e07a5b844446e561e67e2",
    strip_prefix = "rules_go-4c9a52aba0b59511c5646af88d2f93a9c0193647",
    urls = [
        "http://bazel-mirror.storage.googleapis.com/github.com/bazelbuild/rules_go/archive/4c9a52aba0b59511c5646af88d2f93a9c0193647.tar.gz",
        "https://github.com/bazelbuild/rules_go/archive/4c9a52aba0b59511c5646af88d2f93a9c0193647.tar.gz",
    ],
)

load("@io_bazel_rules_go//go:def.bzl", "go_repositories", "go_repository")

go_repositories()

go_repository(
    name = "protobuf",
    commit = "123456789",
    importpath = "github.com/golang/protobuf",
)

go_repository(
    name = "protobuf",
    commit = "ab9f9a6dab164b7d1246e0e688b0ab7b94d8553e",
    importpath = "github.com/golang/protobuf",
)
```

`bazel build @protobuf//...` doesn't complain about the invalid first rule. (although it fails due to having no targets)

### Expected behaviour

Bazel fails and, ideally, reports where the conflicting rules were defined.

### Environment info
* Operating System: Ubuntu 14.04
* Bazel version: 0.5.3

**Comment by ianthehat** [comment]
This is a feature of Bazel.
Repository rules behave differently to normal rules, and this is one of the ways. It's designed to allow a child repository to install default versions of dependencies, and allow a parent to override those defaults, and it's really useful in that regard, so I don't think it's going to go away, but either way, it's a Bazel thing not a go_rules thing, sorry.

**Comment by damonkohler** [comment]
@ianthehat This appears to be inconsistent with "native" repository definitions (e.g. git_repository) which do not allow redefinition after the first load statement (see bazelbuild/bazel#1984). From that perspective, this does appear to be a rules_go thing and perhaps should be documented as a significant difference that requires special consideration? Or perhaps Bazel should reconsider its approach (see #750)?

One suggestion: have overriding repository definitions require a boolean flag "override = True" or they throw an error? This would prevent people from overriding repositories by mistake.

**Comment by drigz** [comment]
On bazel-discuss@, we were recommended to make the child repo only create the rules for the dependencies if they don't already exist (implemented in #750 for some of rules_go's deps).

Given this, perhaps go_repository should issue a warning (or error, if we're not worried about breaking people) when a repository of the same name already exists?

**Comment by drigz** [comment]
@ianthehat Is there any chance of reconsidering this? `go_rules_dependencies` now uses `_maybe` to avoid redefining repositories, so adding a warning would be safe.

This continues to be a pain point for us - for example, I just found that grpc-go wasn't working because http2.Framer had no SetReuseFrames method, even though I had set the newest SHA1 for golang.org/x/net. I had to move the go_repository definition for x/net to before the go_rules_dependencies() invocation.

**Comment by jayconrod** [comment]
I don't think we should change `go_repository` to do anything out of the ordinary for a Skylark repository rule. It should have the same behavior with respect to overriding that other Skylark rules have. If that behavior doesn't make sense, it should be fixed in Bazel itself.

> This continues to be a pain point for us - for example, I just found that grpc-go wasn't working because http2.Framer had no SetReuseFrames method, even though I had set the newest SHA1 for golang.org/x/net. I had to move the go_repository definition for x/net to before the go_rules_dependencies() invocation.

It sounds like `org_golang_x_net` got overridden with an older revision. I'm not sure what happened. We declared an older revision in `go_rules_dependencies` up until a couple days ago, but it's behind `_maybe`, so it shouldn't override any previously declared repo. If you declare it after calling `go_rules_dependencies`, your version should override.

**Comment by drigz** [comment]
@jayconrod, Skylark repository rules fail on redefinition after the first `load` statement, eg for `git_repository`:

`ERROR: .../WORKSPACE:9:1: Cannot redefine repository after any load statement in the WORKSPACE file (for repository 'io_bazel_rules_go')`

Since `go_repository` must be after a `load` statement, it should also fail (or at least warn) on redefinition.

> It sounds like org_golang_x_net got overridden with an older revision.

I had an older version of `go_rules_dependencies`, which was called first, setting an older version of `org_golang_x_net`. Later in the WORKSPACE I had a top-level rule for `org_golang_x_net`, which was silently ignored.

**Comment by jayconrod** [comment]
> Skylark repository rules fail on redefinition after the first load statement, eg for git_repository

Native repository rules fail on redefinition after the first load statement. Skylark repository rules do not. `git_repository` is a native rule, so it has different behavior than `go_repository`. Personally, I don't think that policy makes sense, but it should be changed in Bazel, not rules_go.

> Since go_repository must be after a load statement, it should also fail (or at least warn) on redefinition.

That's not how Skylark rules normally behave. We shouldn't break convention here. If you need that behavior, you can always wrap `go_repository` in a macro like `_maybe`.

**Comment by drigz** [comment]
Thanks for the clarification on Skylark vs native repository rules, I'd misunderstood. You were right that this is an upstream bug, which Damien just filed an hour ago (coincidentally, as far as I can tell): bazelbuild/bazel#3908
